app/src/main/java/com/example/recipeapp/viewmodels/DetailsViewModel.java

```diff
@@ -0,0 +1,147 @@
+import com.example.recipeapp.R;
+import com.example.recipeapp.Retrofit.RecipeApi;
+import com.example.recipeapp.Retrofit.RetrofitClientInstance;
+import com.example.recipeapp.Room.RecipeDatabase;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+public class DetailsViewModel extends ViewModel {
+    private static final String TAG = "DetailsViewModel";
+    public MutableLiveData<List<Ingredient>> showIngredients = new MutableLiveData<>();
+    public MutableLiveData<Integer> bookmarkToast = new MutableLiveData<>();
+    private MutableLiveData<List<StepViewModel>> steps = new MutableLiveData<>();
+
+    private Recipe recipe;
+
+    public DetailsViewModel(Recipe recipe) {
+        this.recipe = recipe;
+        steps.setValue(new ArrayList<>());
+        setDetails();
+        if(recipe.getIngredients() == null) {
+            RecipeApi service = RetrofitClientInstance.getRetrofitInstance().create(RecipeApi.class);
+            Call<Recipe> recipeById = service.getRecipeById(recipe.getId(), BuildConfig.API_KEY);
+            recipeById.enqueue(new Callback<Recipe>() {
+                @Override
+                public void onResponse(Call<Recipe> call, Response<Recipe> response) {
+                    recipe.setAnalyzedInstructions(response.body().getAnalyzedInstructions());
+                    recipe.setIngredients(response.body().getIngredients());
+                }
+
+                @Override
+                public void onFailure(Call<Recipe> call, Throwable t) {
+                    Log.e(TAG, "Something went wrong: " + t);
+                }
+            });
+        }
+    }
+
+    public MutableLiveData<List<StepViewModel>> getSteps() {
+        return steps;
+    }
+
+    private void setDetails() {
+        List<StepViewModel> stepViewModels = new ArrayList<>();
+        for(Step item : recipe.getAnalyzedInstructions()) {
+            stepViewModels.add(new StepViewModel(item.step, item.number));
+        }
+        steps.setValue(stepViewModels);
+    }
+
+    public void showIngredients() {
+        showIngredients.setValue(recipe.getIngredients());
+    }
+
+    public void bookmark() {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        Handler handler = new Handler(Looper.getMainLooper());
+        executor.execute(() -> {
+            RecipeDatabase recipeDatabase = RecipeDatabase.getRecipeDatabase();
+            if(recipe.isBookmarked == null) {
+                recipeDatabase.runInTransaction(() -> {
+                    recipe.isBookmarked = (recipeDatabase.recipeDao().getRecipeById(recipe.id) > 0);
+                    changeBookmark(handler, recipeDatabase);
+                });
+            }
+            else {
+                changeBookmark(handler, recipeDatabase);
+            }
+        });
```

Comment:
nit:Maybe we can probably create repository , so that we do not need to create executor for all the database actions